### DISCUSSION TITLE
PMF Ground Classification Multithreading?

### DISCUSSION THREAD

**Discussion by myerbro3** [r-lidar/lidR#746] [Q&A] [answered]
Hello all, 

I am running a ground classification with the PMF algorithm. 
```
library ("lidR")

is.parallelised(pmf(ws = 5, th = 3))
# FALSE
lasclass <- classify_ground(las, algorithm = pmf(ws = 5, th = 3)) 
# Single Thread
```
A colleague tells me she has been able to run this on multiple threads but I cannot replicate her results, something like: 

`Morphological filter: 1% (8 threads)`

From the docs, I can use `is.parallelised()` and see that the PMF algorithm is not parallelised, confirmed by running the classification I see:

`Morphological filter: 1% (1 threads)`

Our script is the same.  I have used `set_lidr_threads() ` for which `get_lidr_threads()` returns `32`. Could she be mistaken, or am I missing something here? 

Thank you, looking forward to any advice. 
-Brooks

**Comment by Jean-Romain** [comment]
Good question. According to the source code it is not parallelized so `is.parallelised()` is correct.  Why it is not parallelized? I don't know. It could be parallelized, I don't see anything that make it hard to parallelize.

About your colleague, I don't know. I swear there is no parallelization directive in the code.

**Comment by myerbro3** [comment]
Jean-Romain, 

Thank you for your quick response. Is parallelizing this code something that I could do as a user? I am proficient in R but have no experience with writing or editing packages. 

Alternatively, I see in the [docs ](https://r-lidar.github.io/lidRbook/engine.html#engine-options-parallel)that parallelization is made possible through a `LAScatalog` using something like this from 14.7.4: 

```
library(future)
plan(multisession)
opt_chunk_size(ctg) <- 400
opt_chunk_buffer(ctg) <- 40
opt_output_files(ctg) <- paste0(tempdir(), "/{*}_classified")
classified_ctg <- classify_ground(ctg, csf())
```
If I were to use the above code, replacing `csf()` with `pmf()`, should I see an increase in performance compared to processing each .las individually? I have ~280 GB of high resolution UAS .las data (roughly 8 billion points), currently in 113 ~1-3 GB tiles, but these can be re-tiled to whatever size I need. I have an  i9-13900K and 64 GB RAM. 

Best, 
Brooks

**Reply by Jean-Romain** [reply, marked as answer]
> Is parallelizing this code something that I could do as a user?

It is probably a single line of code to change in the c++ source code here https://github.com/r-lidar/lidR/blob/5a5a3a58bbdd1054e2efa54b923ee50c3391157c/src/LAS.cpp#L142

But we must ensure it is really parallelizable first.

> If I were to use the above code, replacing csf() with pmf(), should I see an increase in performance compared to processing each .las individually? 

You will read, classify and write multiple files simultaneously. It will very likely be more performant if done properly; don't use all cores, ensure your computer has enough memory, stuff like that. 64 GB is a lot but lidR is far from optimal in terms of memory usage by design. If you put 6 or 8 cores on it you may compute 4 times faster. As always with parallelization, testing it first is the only solution to measure the actual gain.

**Reply by myerbro3** [reply]
I see, thank you for this very helpful information. I'll give the LAScatalog a shot and see how it goes.

**Reply by Jean-Romain** [reply]
```
plan(multisession)
```
will use all your cores. Use

```
plan(multisession(workers = 6))
```
